### PR TITLE
[FSI] Embedded FSI solver

### PR DESCRIPTION
**Pull request by rubenzorrilla**
New embedded FSI solver valid for both volume and volume-less structures.

### CODE REVIEW

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 1]
```diff
@@ -0,0 +1,665 @@
...
+
+    def _solve_structure(self):
+        # Solve the structure problem
+        self.structure_solver.SolveSolutionStep()
+
+    def _check_FSI_convergence(self, residual_norm):
+        interface_dofs = self._get_partitioned_fsi_utilities().GetInterfaceResidualSize(self._GetStructureInterfaceSubmodelPart())
+        normalised_residual = residual_norm/sqrt(interface_dofs)
+        KratosMultiphysics.Logger.PrintInfo('PartitionedEmbeddedFSIBaseSolver', '\t|res|/sqrt(nDOFS) = ' + str(normalised_residual))
+        if normalised_residual < self.nl_tol:
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]:
```suggestion
        return normalised_residual < self.nl_tol:
```

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 2]
```diff
@@ -0,0 +1,665 @@
...
+            self._get_embedded_skin_utility().InterpolateDiscontinuousMeshVariableToSkin(
+                KratosMultiphysics.PRESSURE, KratosMultiphysics.POSITIVE_FACE_PRESSURE, "positive")
+            self._get_embedded_skin_utility().InterpolateDiscontinuousMeshVariableToSkin(
+                KratosMultiphysics.PRESSURE, KratosMultiphysics.NEGATIVE_FACE_PRESSURE, "negative")
+
+            mapper_params = KratosMultiphysics.Parameters("""{
+                "mapper_type": "nearest_element",
+                "echo_level" : 0
+            }""")
+            mapper = KratosMapping.MapperFactory.CreateMapper(
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py:476]: do you have to recreate it each time?
There would also be `UpdateInterface`

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 3]
```diff
@@ -0,0 +1,665 @@
...
+        fsi_coupling_interface_fluid = fsi_coupling_interface.FSICouplingInterface(
+            self.model,
+            aux_settings)
+
+        KratosMultiphysics.Logger.PrintInfo('PartitionedEmbeddedFSIBaseSolver', 'Fluid FSI coupling interface created')
+
+        return fsi_coupling_interface_fluid
+
+    # This method finds the maximum buffer size between mesh,
+    # fluid and structure solvers and sets it to all the solvers.
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: why do you need it in all solvers?

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 4]
```diff
@@ -0,0 +1,665 @@
...
+        elif (self.level_set_type == "discontinuous"):
+            embedded_skin_utility_skin_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.POSITIVE_FACE_PRESSURE)
+            embedded_skin_utility_skin_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.NEGATIVE_FACE_PRESSURE)
+        else:
+            err_msg = 'Level set type is: \'' + self.level_set_type + '\'. Expected \'continuous\' or \'discontinuous\'.'
+            raise Exception(err_msg)
+        return embedded_skin_utility_skin_model_part
+
+    def _get_structure_interface_model_part_name(self):
+        str_int_list = self.settings["coupling_settings"]["structure_interfaces_list"]
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py:406]: why is it a list if it cannot be more than one?

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 5]
```diff
@@ -0,0 +1,665 @@
...
+            self._embedded_intersections_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.PRESSURE)
+            self._embedded_intersections_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VELOCITY)
+            self._embedded_intersections_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.MESH_VELOCITY)
+            self._embedded_intersections_model_part.AddNodalSolutionStepVariable(KratosMultiphysics.VECTOR_PROJECTED)
+        return self._embedded_intersections_model_part
+
+    def GetOutputVariables(self):
+        pass
+
+    def SaveRestart(self):
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]:
```suggestion

```

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 6]
```diff
@@ -0,0 +1,665 @@
...
+        # Finalize solution step
+        self.fluid_solver.FinalizeSolutionStep()
+        self.structure_solver.FinalizeSolutionStep()
+        self._get_convergence_accelerator().FinalizeSolutionStep()
+
+    def SetEchoLevel(self, structure_echo_level, fluid_echo_level):
+        self.fluid_solver.SetEchoLevel(self, fluid_echo_level)
+        self.structure_solver.SetEchoLevel(self, structure_echo_level)
+
+    def SetTimeStep(self, step):
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: why is this needed?
Doesn't it mess up things?
```suggestion
    def __SetTimeStep(self, step):
```

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py [hunk 7]
```diff
@@ -0,0 +1,665 @@
...
+
+    def Check(self):
+        self.fluid_solver.Check()
+        self.structure_solver.Check()
+
+    #######################################################################
+    ##############          PRIVATE METHODS SECTION          ##############
+    #######################################################################
+
+    def _get_distance_to_skin_process(self, update_distance_process = False):
```
Comment (philbucher) [inline applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]:
```suggestion
    def __GetDistanceToSkinProcess(self, update_distance_process = False):
```

Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: Done for all the "private" methods

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 6 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: Actually I think it is not used in any place.

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: I think it was required in the very first approach of the solver to compute the fixed mesh ALE displacement. With the current implementation I agree that it doesn't make sense at all. I'll try to remove it.

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 4 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py:406]: The GUI allows to select more than one model part as structure interface to consider double-sided surfaces in the body-fitted case. Since there is no embedded FSI GUI yet, I decided to keep the I/O as similar as possible to the body-fitted case.

In any case, this is the very first implementation of the solver. I'm not sure at all if we should allow the user to select more than one model part as interface and internally to merge them in a unique model part.

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 5 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: Both methods have been removed.

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 1 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: Done

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 6 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: Effectively, It wasn't used (removed).

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 3 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py]: Removed. This is not needed anymore.

File: applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py (hunk 2 above)
Comment (rubenzorrilla) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py:476]: The point is that the embedded skin model part is changing its entities (nodes and conditions) each time the level set is updated. Thus I think it is needed to recreate it each time. Am I wrong?

Comment (philbucher) [reply applications/FSIapplication/python_scripts/partitioned_embedded_fsi_base_solver.py:476]: recreating it is certainly not wrong, but might be less efficient than using `UpdateInterface`
=> for now it does the same in the end, so you can leave it as is

**Review by philbucher — DISMISSED** [review]
looks ok to me, nice work :+1:

**Comment by philbucher** [comment]
codacy still has some minor things

**Comment by rubenzorrilla** [comment]
@philbucher done

**Review by philbucher — APPROVED** [review]
:+1:
